sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,612 @@
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return checkpointMark;
+    }
+
+    @Override
+    public byte[] getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public UnboundedMqttSource getCurrentSource() {
+      return source;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write and send a message to a MQTT server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+    @Nullable abstract ConnectionConfiguration connectionConfiguration();
+    @Nullable abstract int qos();
+    @Nullable abstract boolean retained();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration configuration);
+      abstract Builder setQos(int qos);
+      abstract Builder setRetained(boolean retained);
+      abstract Write build();
+    }
+
+    /**
+     * Define MQTT connection configuration used to connect to the MQTT broker.
+     */
+    public Write withConnectionConfiguration(ConnectionConfiguration configuration) {
+      checkArgument(configuration != null,
+          "MqttIO.write().withConnectionConfiguration(configuration) called with null "
+              + "configuration or not called at all");
+      return builder().setConnectionConfiguration(configuration).build();
+    }
+
+    /**
+     * Define the MQTT message quality of service.
+     *
+     * <ul>
+     * <li>Quality of Service 0 (at most once) - a message won't be ack by the receiver or
+     * stored and redelivered by the sender. This is "fire and forget" and provides the same
+     * guarantee as the underlying TCP protocol.</li>
+     *
+     * <li>Quality of Service 1 (at least once) - a message will be delivered at least once to
+     * the receiver. But the message can also be delivered more than once. The sender will
+     * store the message until it gets an ack from the receiver.</li>
+     *
+     * <li>Quality of Service 2 (exactly one) - each message is received only once by the
+     * counterpart. It is the safest and also the slowest quality of service level. The
+     * guarantee is provided by two flows there and back between sender and receiver.</li>
+     * </ul>
+     *
+     * <p>If persistence is not configured, QoS 1 and 2 messages will still be delivered
+     * in the event of a network or server problem as the client will hold state in memory.
+     * If the MQTT client is shutdown or fails and persistence is not configured then
+     * delivery of QoS 1 and 2 messages can not be maintained as client-side state will
+     * be lost.
```

Comment:
I'd like to again ask to prohibit QoS 2 and explicitly mention in the docs that QoS 2 is forbidden because exactly-once publishing can not be implemented within the Beam model due to retries.

Actually, due to retries, QoS 0 will end up being "no guarantees at all" (zero or more times). Document that too.

You may even want to remove `withQoS` and instead have an `enum DeliveryGuarantee { BEST_EFFORT, AT_LEAST_ONCE }`, document it (saying `BEST_EFFORT` uses QoS 0 and will deliver messages zero or more times, and `AT_LEAST_ONCE` uses QoS 1 and will deliver one or more times - leave it at that, and link to official MQTT documentation instead of elaborating more on what QoS is) and have `withDeliveryGuarantee` instead.